File: app_leadership_view/api_client.py
"""
API Client for Knowledge-Weaver Leadership Dashboard
Handles communication with the FastAPI backend
"""
import requests
import streamlit as st
from typing import Dict, List, Any, Optional
import logging

logger = logging.getLogger(__name__)


class APIClient:
    """Client for interacting with Knowledge-Weaver Backend API"""
    
    def __init__(self):
        """Initialize API client with settings from Streamlit secrets"""
        try:
            self.base_url = st.secrets.get("BACKEND_API_URL", "http://localhost:8000")
            self.api_key = st.secrets.get("BACKEND_API_KEY", "dev-secret-key-12345")
        except Exception as e:
            logger.warning(f"Failed to load secrets, using defaults: {e}")
            self.base_url = "http://localhost:8000"
            self.api_key = "dev-secret-key-12345"
        
        self.headers = {
            "X-API-Key": self.api_key,
            "Content-Type": "application/json"
        }
        
        logger.info(f"APIClient initialized with base_url: {self.base_url}")
    
    def _make_request(
        self,
        method: str,
        endpoint: str,
        params: Optional[Dict] = None,
        json_data: Optional[Dict] = None
    ) -> Dict[str, Any]:
        """
        Make HTTP request to backend API
        
        Args:
            method: HTTP method (GET, POST, etc.)
            endpoint: API endpoint path
            params: Query parameters
            json_data: JSON body data
        
        Returns:
            Response data as dictionary
        
        Raises:
            Exception: If request fails
        """
        url = f"{self.base_url}{endpoint}"
        
        try:
            response = requests.request(
                method=method,
                url=url,
                headers=self.headers,
                params=params,
                json=json_data,
                timeout=10
            )
            
            response.raise_for_status()
            return response.json()
        
        except requests.exceptions.Timeout:
            logger.error(f"Request timeout for {endpoint}")
            raise Exception("Request timeout - backend may be slow or unavailable")
        
        except requests.exceptions.ConnectionError:
            logger.error(f"Connection error for {endpoint}")
            raise Exception("Cannot connect to backend API - please check if it's running")
        
        except requests.exceptions.HTTPError as e:
            logger.error(f"HTTP error for {endpoint}: {e}")
            if response.status_code == 401:
                raise Exception("Authentication failed - invalid API key")
            elif response.status_code == 404:
                raise Exception("Endpoint not found")
            elif response.status_code == 500:
                raise Exception("Backend server error")
            else:
                raise Exception(f"HTTP {response.status_code}: {str(e)}")
        
        except Exception as e:
            logger.error(f"Unexpected error for {endpoint}: {e}")
            raise Exception(f"Request failed: {str(e)}")
    
    def fetch_query_logs(
        self,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        limit: int = 100
    ) -> List[Dict[str, Any]]:
        """
        Fetch query logs from backend for analytics
        
        Args:
            start_date: Start date filter (ISO format)
            end_date: End date filter (ISO format)
            limit: Maximum number of logs to return
        
        Returns:
            List of query log entries
        """
        params = {"limit": limit}
        if start_date:
            params["start_date"] = start_date
        if end_date:
            params["end_date"] = end_date
        
        try:
            result = self._make_request("GET", "/api/v1/metrics/queries", params=params)
            return result.get("logs", [])
        except Exception as e:
            logger.error(f"Failed to fetch query logs: {e}")
            st.error(f"Failed to fetch query logs: {str(e)}")
            return []
    
    def fetch_unanswered_questions(self) -> List[Dict[str, Any]]:
        """
        Fetch unanswered questions from backend
        
        Returns:
            List of unanswered question entries
        """
        try:
            result = self._make_request("GET", "/api/v1/metrics/unanswered")
            return result.get("unanswered_questions", [])
        except Exception as e:
            logger.error(f"Failed to fetch unanswered questions: {e}")
            st.error(f"Failed to fetch unanswered questions: {str(e)}")
            return []
    
    def test_connection(self) -> bool:
        """
        Test connection to backend API
        
        Returns:
            True if connection successful, False otherwise
        """
        try:
            result = self._make_request("GET", "/api/v1/health")
            return result.get("status") == "healthy"
        except Exception as e:
            logger.error(f"Connection test failed: {e}")
            return False
    
    def get_health_status(self) -> Dict[str, Any]:
        """
        Get backend health status
        
        Returns:
            Health status dictionary
        """
        try:
            return self._make_request("GET", "/api/v1/health")
        except Exception as e:
            logger.error(f"Failed to get health status: {e}")
            return {"status": "error", "message": str(e)}

    def fetch_recent_knowledge(self, limit: int = 10) -> List[Dict[str, Any]]:
        """
        Fetch recent knowledge entries
        
        Args:
            limit: Maximum number of entries to return
            
        Returns:
            List of knowledge entries
        """
        try:
            return self._make_request("GET", "/api/v1/knowledge/recent", params={"limit": limit})
        except Exception as e:
            logger.error(f"Failed to fetch recent knowledge: {e}")
            return []

    def delete_knowledge_entry(self, entry_id: str) -> bool:
        """Delete a knowledge entry"""
        try:
            response = requests.delete(
                f"{self.base_url}/knowledge/{entry_id}",
                headers=self.headers,
                timeout=10
            )
            return response.status_code == 200
        except Exception as e:
            print(f"Error deleting entry: {e}")
            return False

    def update_knowledge_entry(self, entry_id: str, updates: dict) -> bool:
        """Update a knowledge entry"""
        try:
            response = requests.patch(
                f"{self.base_url}/knowledge/{entry_id}",
                headers=self.headers,
                json=updates,
                timeout=10
            )
            return response.status_code == 200
        except Exception as e:
            print(f"Error updating entry: {e}")
            return False

    def fetch_dashboard_metrics(self) -> dict:
        """Fetch aggregated dashboard metrics"""
        try:
            result = self._make_request("GET", "/api/v1/metrics/dashboard")
            return result
        except Exception as e:
            print(f"Error fetching dashboard metrics: {e}")
            return None

File: app_agent_view/scripts/api_client.js

/**
 * Backend API Client for Knowledge-Weaver Chrome Extension
 * Handles communication with the FastAPI backend
 */

class BackendAPIClient {
    constructor() {
        this.baseURL = null;
        this.apiKey = null;
        this.maxRetries = 2;
        this.timeout = 60000; // 60 seconds timeout for AI operations
    }

    /**
     * Load settings from Chrome storage
     */
    async loadSettings() {
        return new Promise((resolve, reject) => {
            chrome.storage.local.get(['apiEndpoint', 'apiKey'], (result) => {
                if (chrome.runtime.lastError) {
                    reject(new Error('Failed to load settings from storage'));
                    return;
                }

                this.baseURL = result.apiEndpoint || 'http://localhost:8000';
                this.apiKey = result.apiKey || 'dev-secret-key-12345';

                console.log('API Client settings loaded:', {
                    baseURL: this.baseURL,
                    hasApiKey: !!this.apiKey
                });

                resolve();
            });
        });
    }

    /**
     * Query the knowledge base with natural language
     * @param {string} query - Natural language query
     * @param {boolean} verifiedOnly - Filter for verified content only
     * @returns {Promise<Object>} Query results
     */
    async queryKnowledgeBase(query, verifiedOnly = false) {
        if (!query || !query.trim()) {
            throw new Error('Query cannot be empty');
        }

        // Load settings if not already loaded
        if (!this.baseURL || !this.apiKey) {
            await this.loadSettings();
        }

        console.log('Querying knowledge base:', query.substring(0, 100) + '...', 'Verified Only:', verifiedOnly);

        // Retry logic with exponential backoff
        let lastError = null;

        for (let attempt = 0; attempt < this.maxRetries; attempt++) {
            try {
                const result = await this.makeRequest(query, verifiedOnly, attempt);
                return result;
            } catch (error) {
                lastError = error;
                console.warn(`Query attempt ${attempt + 1}/${this.maxRetries} failed:`, error.message);

                // If not the last attempt, wait before retrying
                if (attempt < this.maxRetries - 1) {
                    const backoffTime = Math.pow(2, attempt) * 1000; // 1s, 2s
                    console.log(`Retrying in ${backoffTime}ms...`);
                    await this.sleep(backoffTime);
                }
            }
        }

        // All retries failed
        console.error('All retry attempts failed');
        throw lastError;
    }

    /**
     * Make the actual HTTP request to the backend
     * @param {string} query - Query text
     * @param {boolean} verifiedOnly - Filter for verified content only
     * @param {number} attemptNumber - Current attempt number
     * @returns {Promise<Object>} API response
     */
    async makeRequest(query, verifiedOnly, attemptNumber) {
        const url = `${this.baseURL}/api/v1/knowledge/query`;

        console.log(`Making request (attempt ${attemptNumber + 1}):`, url);

        // Create abort controller for timeout
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);

        try {
            const response = await fetch(url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'X-API-Key': this.apiKey
                },
                body: JSON.stringify({
                    query: query,
                    verified_only: verifiedOnly
                }),
                signal: controller.signal
            });

            clearTimeout(timeoutId);

            // Handle HTTP errors
            if (!response.ok) {
                await this.handleHTTPError(response);
            }

            const data = await response.json();
            console.log('Query successful:', data);

            return data;

        } catch (error) {
            clearTimeout(timeoutId);

            if (error.name === 'AbortError') {
                throw new Error('Request timeout (30 seconds exceeded)');
            }

            if (error.message.includes('Failed to fetch')) {
                throw new Error('Network error: Unable to reach backend API. Please check if the backend is running.');
            }

            throw error;
        }
    }

    /**
     * Handle HTTP error responses
     * @param {Response} response - Fetch response object
     */
    async handleHTTPError(response) {
        let errorMessage = `HTTP ${response.status}: ${response.statusText}`;

        try {
            const errorData = await response.json();
            if (errorData.detail) {
                errorMessage = errorData.detail;
            }
        } catch (e) {
            // Could not parse error response
        }

        if (response.status === 401) {
            throw new Error('Authentication failed: Invalid API key');
        } else if (response.status === 400) {
            throw new Error(`Bad request: ${errorMessage}`);
        } else if (response.status === 500) {
            throw new Error('Server error: The backend encountered an error');
        } else if (response.status === 503) {
            throw new Error('Service unavailable: Backend services not initialized');
        } else {
            throw new Error(errorMessage);
        }
    }

    /**
     * Sleep utility for retry backoff
     * @param {number} ms - Milliseconds to sleep
     */
    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    /**
     * Analyze content for category, tags, and summary
     * @param {Object} payload - { text, url, screenshot, timestamp }
     * @returns {Promise<Object>} - Analysis result
     */
    async analyzeContent(payload) {
        // Load settings if not already loaded
        if (!this.baseURL || !this.apiKey) {
            await this.loadSettings();
        }

        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), this.timeout);

        try {
            const response = await fetch(`${this.baseURL}/api/v1/analyze`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'X-API-Key': this.apiKey
                },
                body: JSON.stringify(payload),
                signal: controller.signal
            });

            clearTimeout(timeoutId);

            if (!response.ok) {
                const errorData = await response.json().catch(() => ({}));
                throw new Error(errorData.message || `Analysis failed: ${response.status}`);
            }

            return await response.json();
        } catch (error) {
            clearTimeout(timeoutId);
            if (error.name === 'AbortError') {
                throw new Error('Request timed out. The AI analysis is taking longer than expected.');
            }
            throw error;
        }
    }

    /**
     * Ingest knowledge manually
     * @param {Object} payload - Ingestion payload {text, screenshot, url, timestamp}
     * @returns {Promise<Object>} API response
     */
    async ingestKnowledge(payload) {
        if (!payload || !payload.text) {
            throw new Error('Payload must contain text');
        }

        // Load settings if not already loaded
        if (!this.baseURL || !this.apiKey) {
            await this.loadSettings();
        }

        const url = `${this.baseURL}/api/v1/ingest`;
        console.log('Ingesting knowledge:', url);

        try {
            const response = await fetch(url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'X-API-Key': this.apiKey
                },
                body: JSON.stringify(payload)
            });

            if (!response.ok) {
                await this.handleHTTPError(response);
            }

            const data = await response.json();
            console.log('Ingestion successful:', data);
            return data;

        } catch (error) {
            console.error('Ingestion failed:', error);
            if (error.message.includes('Failed to fetch')) {
                throw new Error('Network error: Unable to reach backend API.');
            }
            throw error;
        }
    }

    /**
     * Test connection to backend
     * @returns {Promise<boolean>} True if connection successful
     */
    async testConnection() {
        try {
            await this.loadSettings();

            const url = `${this.baseURL}/api/v1/health`;
            const response = await fetch(url, {
                method: 'GET',
                signal: AbortSignal.timeout(3000)
            });

            if (!response.ok) {
                return false;
            }

            const data = await response.json();
            console.log('Backend health check:', data);

            return data.status === 'healthy';

        } catch (error) {
            console.error('Connection test failed:', error);
            return false;
        }
    }
}

// Export for use in content script
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { BackendAPIClient };
}
